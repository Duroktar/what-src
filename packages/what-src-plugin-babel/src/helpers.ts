import template from '@babel/template'
import { JSXOpeningElement, JSXElement } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { defaultOptions } from './options'
import { toCamelCase, getIn, isNullOrUndefined } from './utils'
import * as GIT from './git'
import * as T from './types'

/**
 * returns any default options merged with all passed plugin options
 *
 * @param {T.WhatSrcPluginOptions} options
 * @param {*} [defaults=defaultOptions]
 * @returns {Required<T.WhatSrcPluginOptions>}
 */
export const getAllPluginOptions = (
  options: T.WhatSrcPluginOptions,
  defaults = defaultOptions,
): Required<T.WhatSrcPluginOptions> => {
  return { ...defaults, ...options }
}

/**
 * utility function for resolving file urls to their git remote
 * counterparts.
 */
export const gitUrlResolver = (() => {
  const remoteUrl = GIT.getGitRemoteUrl()
  const branch = GIT.getGitBranchName()
  return function getRemoteFileUrl(filepath: string) {
    const relativeFileUrl = GIT.getRelativeGitFileUrl(filepath)
    const options = { branch, filepath: relativeFileUrl }
    return generateGitFileUrl(remoteUrl, options)
  }
})()

/**
 * resolves the github url for the current filename when the useRemote option is
 * set
 *
 * @param {string} filename
 * @param {Required<T.WhatSrcPluginOptions>} options
 * @param {(str: string) => string} resolver
 */
export const getRemoteFilenameIfSet = (
  filename: string,
  options: Required<T.WhatSrcPluginOptions>,
) => options.useRemote ? gitUrlResolver(filename) : filename

/**
 * generates the github url for the current git branch and filepath
 *
 * @param {string} remoteUrl
 * @param {{
 *   branch: string;
 *   filepath: string;
 * }} opts
 * @returns
 */
export const generateGitFileUrl = (remoteUrl: string, opts: {
  branch: string;
  filepath: string;
}) => {
  if (isNullOrUndefined(remoteUrl) || !remoteUrl.endsWith('.git')) {
    throw new Error(`${remoteUrl} is not a valid remote url.`)
  }
  return `${remoteUrl.slice(0, -4)}/blob/${opts.branch}/${opts.filepath}`
}

/**
 * returns the generated prperty name for a data attribute string
 *
 * @param {string} spinalName
 * @returns
 */
export const parseDataTag = (spinalName: string) => {
  return toCamelCase(spinalName.slice('data-'.length))
}

/**
 * generate the ast for the main clickhandler
 *
 * @param {T.BabelPluginContext['types']} t
 * @param {Required<T.WhatSrcPluginOptions>} options
 * @param {({ [key: string]: string | undefined })} cache
 * @returns
 */
export const generateClickHandlerAst = (
  t: T.BabelPluginContext['types'],
  options: Required<T.WhatSrcPluginOptions>,
  cache: { [key: string]: string | undefined }
) => {
  return clickHandlerBuilder({
    globalCacheKey: t.stringLiteral(options.globalCacheKey),
    serverUrl: t.stringLiteral(options.serverUrl),
    cache: t.stringLiteral(JSON.stringify(cache)),
    dataTag: t.stringLiteral(parseDataTag(options.dataTag)),
    stopPropagation: t.booleanLiteral(options.stopPropagation),
    preventDefault: t.booleanLiteral(options.preventDefault),
    useRemote: t.booleanLiteral(options.useRemote),
  })
}

const clickHandlerBuilder = template.statement(`
  try {
    const cache = JSON.parse(%%cache%%)
    window[%%globalCacheKey%%] = function (e) {
      if (e.metaKey) {
        const dataset = cache[e.path[0].dataset[%%dataTag%%]]
        if (%%stopPropagation%%) e.stopPropagation()
        if (%%preventDefault%%) e.preventDefault()
        if (typeof dataset === 'undefined') return
        if (%%useRemote%%) {
          const {remoteUrl} = JSON.parse(dataset)
          console.log('Opening', remoteUrl, 'in browser tab')
          return window.open(remoteUrl, '_blank')
        }
        const xhr = new XMLHttpRequest()
        xhr.open('POST', %%serverUrl%%, true)
        xhr.setRequestHeader('Content-type', 'application/json')
        xhr.send(dataset)
      }
    }
    window.document.removeEventListener('click', window[%%globalCacheKey%%])
    window.document.addEventListener('click', window[%%globalCacheKey%%])
  } catch {}
`)

/**
 * generate an ast for the configured data attribute tag
 *
 * @param {T.BabelPluginContext['types']} t
 * @param {Required<T.WhatSrcPluginOptions>} options
 * @param {string} nextId
 * @returns
 */
export const generateAttribute = (
  t: T.BabelPluginContext['types'],
  options: Required<T.WhatSrcPluginOptions>,
  nextId: string
) => {
  return t.jsxAttribute(
    t.jsxIdentifier(options.dataTag),
    t.stringLiteral(nextId)
  )
}

/**
 * generates the metadata object for a given nodepath
 *
 * @param {NodePath<JSXElement>} path
 * @param {string} filename
 * @returns
 */
export const generateJsxMetaData = (
  path: NodePath<JSXElement>,
  filename: string,
) => {
  const meta = getIn('node.openingElement.loc.start', path)
  const metaData = {
    filename: filename,
    line: meta.line,
    column: meta.column + 1,
    remoteUrl: '',
  }
  metaData.remoteUrl =
    `${filename}#L${metaData.line}`
  return metaData
}

/**
 * used to determine if an openingElement is a React.Fragment
 *
 * @param {JSXOpeningElement} openingElement
 * @returns
 */
export const isFragment = (openingElement: JSXOpeningElement) => {
  return getIn('name.property.name', openingElement) === 'Fragment'
}
