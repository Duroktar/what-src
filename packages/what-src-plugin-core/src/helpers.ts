import { toCamelCase, isNullOrUndefined, λIf } from '@what-src/utils'
import { defaultOptions } from './options'
import * as GIT from './git'
import * as T from './types'
import { STATS_URL } from './constants'

/**
 * returns any default options merged with all passed plugin options
 *
 * @param {T.WhatSrcPluginOptions} options
 * @param {*} [defaults=defaultOptions]
 * @returns {T.WhatSrcPluginOptions}
 */
export const mergePluginOptions = (
  options: T.WhatSrcPluginOptions,
  defaults = defaultOptions
): T.WhatSrcConfiguration => {
  return { ...defaults, ...options }
}

/**
 * stateful utility function for resolving file urls to their git remote
 * counterparts.
 *
 * > TODO: This should be replaced with
 * >   [hosted-git-info](https://www.npmjs.com/package/hosted-git-info).
 *
 * @param {string} filepath
 * @returns {string} The resolved git url
 */
export const gitUrlResolver = (() => {
  const remoteUrl = GIT.getGitRemoteOriginUrl()
  const branch = GIT.getGitBranchName()
  return function getRemoteFileUrl(filepath: string) {
    const relativeFileUrl = GIT.getGitFileFullName(filepath)
    const options = { branch, filepath: relativeFileUrl }
    return generateGitFileUrl(remoteUrl, options)
  }
})()

/**
 * resolves the github url for the current filename when the useRemote option is
 * set
 *
 * @param {string} filename
 * @param {T.WhatSrcConfiguration} options
 * @param {(str: string) => string} resolver
 */
export const getRemoteFilenameIfSet = (
  filename: string,
  options: T.WhatSrcConfiguration
) => options.useRemote ? gitUrlResolver(filename) : filename

/**
 * generates the github url for the current git branch and filepath
 *
 * @param {string} remoteUrl
 * @param {{
 *   branch: string;
 *   filepath: string;
 * }} opts
 * @returns A valid git resource url. Throws an error otherwise.
 */
export const generateGitFileUrl = (remoteUrl: string, opts: {
  branch: string;
  filepath: string;
}) => {
  if (isNullOrUndefined(remoteUrl) || !remoteUrl.trim().endsWith('.git')) {
    throw new Error(`${remoteUrl} is not a valid remote url.`)
  }
  return `${remoteUrl.slice(0, -4)}/blob/${opts.branch}/${opts.filepath}`
}

/**
 * returns the generated prperty name for a data attribute string
 *
 * @param {string} spinalName
 * @returns
 */
export const parseDataTag = (spinalName: string) => {
  return toCamelCase(spinalName.slice('data-'.length))
}

/**
 * generate the ast for the main clickhandler
 *
 * @param {T.WhatSrcConfiguration} options
 * @param {({ [key: string]: string | undefined })} cache
 * @returns
 */
export const generateClickHandlerRawString = (
  options: T.WhatSrcConfiguration,
  cache: { [key: string]: string | undefined }
) => {
  return ((o = {
    ...options,
    dataTag: parseDataTag(options.dataTag),
    whatSrcStatsUrl: process.env.STATS_URL || STATS_URL,
  }) => `
    const __whatSrcCallback01101011 = (() => {
      try {
        const cache = (${JSON.stringify(cache, null, 3)})
        window["${o.globalCacheKey}"] = function (e) {
          if (e.metaKey) {
            const tag = cache[e.path[0].dataset["${o.dataTag}"]];
            if (tag === undefined) throw new Error('Scott.. Ite derped again.');
            const dataset = JSON.parse(tag);
            ${λIf(o.stopPropagation, 'e.stopPropagation();', '')}
            ${λIf(o.preventDefault, 'e.preventDefault();', '')}
            if (typeof dataset === 'undefined') return;
            ${λIf(o.useRemote, {
              value: 'window.open(dataset.remoteUrl, "_blank");',
              otherwise: `
                const xhr = new XMLHttpRequest();
                xhr.open('POST', "${o.serverUrl}", true);
                xhr.setRequestHeader('Content-type', 'application/json');
                xhr.send(JSON.stringify({ ...dataset, basedir: cache.__basedir }));
              `,
            })}
            ${λIf(o.enableXkcdMode, `
              const xhr = new XMLHttpRequest();
              xhr.open('POST', "${o.whatSrcStatsUrl}", true);
              xhr.send();
            `, '')}
          }
        }
        window.document.removeEventListener('click', window["${o.globalCacheKey}"])
        window.document.addEventListener('click', window["${o.globalCacheKey}"])
        return true
      } catch {
        return false
      }
    })();

    export default __whatSrcCallback01101011
  `)()
}

/**
 * generates the relevant metadata object
 *
 * @param {meta} {
 *   col,
 *   filename,
 *   line,
 * }
 * @returns
 */
export const generateJsxMetaData = (location: T.SourceLocationFullStart) => {
  const { col, filename, line } = location
  return {
    filename: filename,
    line: line,
    column: col,
    remoteUrl: `${filename}#L${line}`,
  }
}
