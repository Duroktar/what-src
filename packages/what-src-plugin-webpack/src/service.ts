#!/usr/bin/env node

require('@what-src/express')
