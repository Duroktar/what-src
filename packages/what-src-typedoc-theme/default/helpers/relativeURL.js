const BASE_URL = '/doc/'

module.exports = {
  relativeURL: function(url, options) {
    // console.log('hhhsdlfjasldsfdhajlsdfs')
    return BASE_URL + url
  },
}
