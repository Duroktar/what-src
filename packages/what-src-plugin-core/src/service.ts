import * as ts from 'typescript'
import { join, dirname } from 'path'
import { withHooks, isNullOrUndefined } from '@what-src/utils'
import { defaultOptions, defaultCache } from './options'
import * as H from './helpers'
import * as T from './types'
import { mkdirSync } from 'fs'

/**
 * Factory function for the WhatSrcService class.
 *
 * @param {"WhatSrcService"} { options, basedir, cache = defaultCache }
 * @returns new what-src service instance
 */
export const getService = ({ options, ...args }: T.ServiceOptions = {}) => {
  return new WhatSrcService(options, args.basedir, args.cache)
}

/**
 * The primary @what-src/plugin-core service class.
 *
 * @export
 * @class WhatSrcService
 */
export class WhatSrcService {
  /**
   * full set of what-src configuration options
   *
   * @type {WS.WhatSrcConfiguration}
   * @memberof WhatSrcBabelPlugin
   */
  public options: T.WhatSrcConfiguration

  /**
   * incrementing value used for element tags
   *
   * @private
   * @type {number}
   * @memberof Resolver
   */
  private nextId: number = 1

  /**
   * list of element tag names to ignore
   *
   * @private
   * @type {Set<string>}
   * @memberof WhatSrcService
   */
  private blockedTags!: Set<string>

  /**
   * the cache *theme song plays
   *
   * @private
   * @type {T.WhatSrcCache}
   * @memberof WhatSrcService
   */
  private _cache!: T.WhatSrcCache

  /**
   * Creates an instance of WhatSrcService.
   *
   * @param {T.WhatSrcPluginOptions} [options={}]
   * @param {string} [basedir='']
   * @param {{}} [cache={}]
   * @memberof WhatSrcService
   */
  constructor(
    options: T.WhatSrcPluginOptions = {},
    basedir: string = '',
    cache: {} = {}
  ) {
    this.options = H.mergePluginOptions(options, defaultOptions)
    this.blockedTags = new Set(this.options.blacklistedTags)
    this._cache = this.initializeCache(basedir, cache)
  }

  /**
   * Initialize the cache with system/user settings and basedir.
   *
   * @private
   * @memberof WhatSrcService
   */
  private initializeCache = (basedir: string, cache: {}) => {
    const __basedir = this.getBaseDirForCache(basedir)
    return { ...defaultCache, ...cache, __basedir }
  }

  /**
   * Compile and emit the cache file to the provided location.
   *
   * @memberof WhatSrcService
   */
  public emit = (location: string) => {
    const source = H.generateClickHandlerRawString(this.options, this._cache)

    // using ts here cause it's got builtin filesystem helpers
    const result = ts.transpileModule(source, {
      compilerOptions: { module: ts.ModuleKind.CommonJS },
    })

    // create the necessary caching directories and paths
    const host = ts.createCompilerHost(this.options)
    if (this.options.createCacheDir && !host.directoryExists!(dirname(location))) {
      mkdirSync(dirname(location), { recursive: true })
    }

    // useful mostly for CI but we can skip the actual write part if needed
    if (this.options.createCacheFile) {
      const transpiledSource = H.compileSource(result.outputText)
      host.writeFile(location, transpiledSource, false, errMsg => {
        throw new Error(
          '[@what-src/core] Error writing file.. ' +
          'Make sure the path is absolute! ->' + errMsg)
      })
    }
  }

  /**
   * caches a location object and returns a retrieval key.
   *
   * @memberof WhatSrcService
   */
  public cache = (loc: T.SourceLocationStart, sourcefile: string) => {
    const remoteFn = H.getRemoteFilenameOrNull(sourcefile, this.options)

    let filename = remoteFn || sourcefile

    if (filename.startsWith(this._cache.__basedir)) {
      filename = filename.slice(this._cache.__basedir.length)
    }

    const metaData = H.generateJsxMetaData(
      { ...loc, filename },
      this.options.useRemote,
    )

    return withHooks(() => this.nextId.toString(), {
      after: () => this.setCache(JSON.stringify(metaData)),
    }).result
  }

  /**
   * ignore any explicitly blacklisted tags. the blacklist is converted to a set
   * during class construction to be used for fast lookups
   *
   * @param {ts.JsxElement} node
   * @returns
   * @memberof WhatSrcTsTransformer
   */
  public tagIsBlacklisted(tagname: string) {
    return this.blockedTags.has(tagname)
  }

  /**
   * get cache file relative import path
   *
   * @returns
   * @memberof WhatSrcTsTransformer
   */
  public getCacheFileImport() {
    return isNullOrUndefined(this.options.cacheRequireOverride)
      ? '@what-src/runtime-cache'
      : this.options.cacheRequireOverride
  }

  /**
   * get the absolute path to the cache file
   *
   * @returns
   * @memberof WhatSrcTsTransformer
   */
  public getCacheFileLocation(rootDir: string, ft = '.js') {
    const path = ['node_modules', '@what-src', 'runtime-cache']
    return join(rootDir, ...path, 'index' + ft)
  }

  /**
   * resolves the base dir for the cache depending on if
   * useRemote is set. the path can be completely overriden
   * by setting the baseDirOverride options
   *
   * @param {string} basedir
   * @returns
   * @memberof WhatSrcService
   */
  public getBaseDirForCache(basedir: string) {
    if (this.options.baseDirOverride) return this.options.baseDirOverride
    else if (this.options.useRemote) return H.getGitRemoteBaseDir()
    else return basedir
  }

  /**
   * returns the source file location cache
   *
   * @memberof WhatSrcService
   */
  public getCache = (): T.WhatSrcCache => {
    return this._cache
  }

  /**
   *  sets the next cache value then increments the key
   *
   * @memberof WhatSrcService
   */
  public setCache = (metaData: string) => {
    this._cache[this.nextId++] = metaData
  }

  /**
   * get the basedir as set in the cache
   *
   * @readonly
   * @memberof WhatSrcService
   */
  public get basedir() {
    return this._cache.__basedir
  }
}
