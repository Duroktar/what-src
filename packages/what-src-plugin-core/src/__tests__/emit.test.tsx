import * as ts from 'typescript'
import { getService } from '../service'
import * as H from '../helpers'

const service = getService({
  basedir: '/home/user/app/packages/',
  cache: {},
  options: {},
})

it('emits a cache file', () => {
  const source = H.generateClickHandlerRawString(
    service.options,
    service.getSourceCache(),
  )

  const result = ts.transpileModule(source, {
    compilerOptions: { module: ts.ModuleKind.CommonJS },
  })

  expect<string>(result.outputText).toMatchSnapshot()
})
