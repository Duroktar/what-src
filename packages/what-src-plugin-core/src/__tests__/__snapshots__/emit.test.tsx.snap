// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits a cache file 1`] = `
"\\"use strict\\";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var __whatSrcCallback01101011 = (function () {
    try {
        var cache_1 = ({
            \\"__basedir\\": \\"/home/user/app/packages/\\"
        });
        window[\\"__what-src-global-callback-key\\"] = function (e) {
            if (e.metaKey) {
                var tag = cache_1[e.path[0].dataset[\\"whatSrc\\"]];
                if (tag === undefined)
                    throw new Error('Scott.. Ite derped again.');
                var dataset = JSON.parse(tag);
                e.stopPropagation();
                e.preventDefault();
                if (typeof dataset === 'undefined')
                    return;
                var xhr = new XMLHttpRequest();
                xhr.open('POST', \\"http://localhost:8018/__what_src\\", true);
                xhr.setRequestHeader('Content-type', 'application/json');
                xhr.send(JSON.stringify(__assign(__assign({}, dataset), { basedir: cache_1.__basedir })));
            }
        };
        window.document.removeEventListener('click', window[\\"__what-src-global-callback-key\\"]);
        window.document.addEventListener('click', window[\\"__what-src-global-callback-key\\"]);
        return true;
    }
    catch (_a) {
        return false;
    }
})();
exports.default = __whatSrcCallback01101011;
"
`;
